src/compute/src/sink/s3_oneshot.rs

```diff
@@ -27,18 +32,116 @@ where
-        _sink: &ComputeSinkDesc<CollectionMetadata>,
-        _sink_id: GlobalId,
+        compute_state: &mut crate::compute_state::ComputeState,
+        sink: &ComputeSinkDesc<CollectionMetadata>,
+        sink_id: GlobalId,
         _as_of: Antichain<Timestamp>,
-        _sinked_collection: Collection<G, Row, Diff>,
-        _err_collection: Collection<G, DataflowError, Diff>,
+        sinked_collection: Collection<G, Row, Diff>,
+        err_collection: Collection<G, DataflowError, Diff>,
     ) -> Option<Rc<dyn Any>>
     where
         G: Scope<Timestamp = Timestamp>,
     {
-        info!("Starting s3 sink dataflow to write data to {}", self.prefix);
-        // TODO(mouli): implement this in subsequent PRs
-        None
+        // An encapsulation of the copy to response protocol.
+        // Used to send rows and errors if this fails.
+        let response_protocol_handle = Rc::new(RefCell::new(Some(ResponseProtocol {
+            sink_id,
+            response_buffer: Some(Rc::clone(&compute_state.copy_to_response_buffer)),
+        })));
+        let response_protocol_weak = Rc::downgrade(&response_protocol_handle);
+
+        copy_to(
+            sinked_collection,
+            err_collection,
+            sink_id,
+            sink.up_to.clone(),
+            response_protocol_handle,
+        );
+
+        Some(Rc::new(scopeguard::guard((), move |_| {
+            if let Some(protocol_handle) = response_protocol_weak.upgrade() {
+                std::mem::drop(protocol_handle.borrow_mut().take())
+            }
+        })))
+    }
+}
+
+fn copy_to<G>(
+    sinked_collection: Collection<G, Row, Diff>,
+    err_collection: Collection<G, DataflowError, Diff>,
+    sink_id: GlobalId,
+    up_to: Antichain<G::Timestamp>,
+    response_protocol_handle: Rc<RefCell<Option<ResponseProtocol>>>,
+) where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let name = format!("s3-oneshot-{}", sink_id);
+    let mut op = OperatorBuilder::new(name, sinked_collection.scope());
+    let mut ok_input = op.new_input(&sinked_collection.inner, Pipeline);
+    let mut err_input = op.new_input(&err_collection.inner, Pipeline);
+
+    op.build(|_cap| {
+        let mut row_count: u64 = 0;
+        let mut finished = false;
+        let mut ok_buf = Default::default();
+
+        move |frontiers| {
+            if finished {
+                // Drain the inputs, to avoid the operator being constantly rescheduled
+                ok_input.for_each(|_, _| {});
+                err_input.for_each(|_, _| {});
+                return;
+            }
+
+            let mut frontier = Antichain::new();
+            for input_frontier in frontiers {
+                frontier.extend(input_frontier.frontier().iter().copied());
+            }
+
+            ok_input.for_each(|_, data| {
+                data.swap(&mut ok_buf);
+                let count: u64 = ok_buf.drain(..).len().try_into().expect("usize into u64");
```

Comment:
You'll need to filter data here for your up-to frontier, because there is no guarantee that a batch at time $t$ will only contain data at time $t$, but instead the guarantee is that forall data's time $t_d: t \leq t_d$. (This independent of what the system does and good habit for defensive programming.)

Reply:
Ack. I will clean this up in subsequent PRs where I will implement this sink properly. Currently I just needed the sink dataflow to return _some_ value for the response protocol to be defined.